x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java

```diff
@@ -542,6 +551,33 @@ public static RoleDescriptor parsePrivilegesCheck(String description, BytesRefer
                 .build();
     }
 
+    private static void checkIfExceptFieldsIsSubsetOfGrantedFields(String roleName, String[] grantedFields, String[] deniedFields) {
+        Automaton grantedFieldsAutomaton;
+        if (grantedFields == null || Arrays.stream(grantedFields).anyMatch(Regex::isMatchAllPattern)) {
+            grantedFieldsAutomaton = Automatons.MATCH_ALL;
+        } else {
+            // an automaton that includes metadata fields, including join fields created by the _parent field such
+            // as _parent#type
+            Automaton metaFieldsAutomaton = Operations.concatenate(Automata.makeChar('_'), Automata.makeAnyString());
+            grantedFieldsAutomaton = Operations.union(Automatons.patterns(grantedFields), metaFieldsAutomaton);
+        }
+        Automaton deniedFieldsAutomaton;
+        if (deniedFields == null || deniedFields.length == 0) {
+            deniedFieldsAutomaton = Automatons.EMPTY;
+        } else {
+            deniedFieldsAutomaton = Automatons.patterns(deniedFields);
+        }
+        grantedFieldsAutomaton = MinimizationOperations.minimize(grantedFieldsAutomaton,
+            Operations.DEFAULT_MAX_DETERMINIZED_STATES);
+        deniedFieldsAutomaton = MinimizationOperations.minimize(deniedFieldsAutomaton,
+            Operations.DEFAULT_MAX_DETERMINIZED_STATES);
+        if (subsetOf(deniedFieldsAutomaton, grantedFieldsAutomaton) == false) {
+            throw new ElasticsearchParseException("failed to parse indices privileges for role [{}]. [{}] field values [{}] must be a " +
+                "subset of [{}] field values [{}]", roleName, Fields.EXCEPT_FIELDS, Strings.arrayToCommaDelimitedString(deniedFields),
+                Fields.GRANT_FIELDS, Strings.arrayToCommaDelimitedString(grantedFields));
+        }
```

Comment:
Can we find a way to unify (some of) this with what's already in `FieldPermissions`. It doesn't seem right to have to checks for the same requirement implemented separately.